<launch>
	<!-- Launch file for Livox Mid360 LiDAR -->
	<arg name="rviz" default="true" />
	<!-- move_base 发布速度控制话题（默认 /cmd_vel），可按需改成 /cmd_vel_nav 等 -->
	<arg name="cmd_vel_topic" default="/cmd_vel_nav" />
	<rosparam command="load" file="$(find faster_lio)/config/mid360_loc.yaml" />

	<param name="runtime_pos_log_enable" type="bool" value="false" />
	<node pkg="faster_lio" type="run_mapping_online" name="laserMapping" output="screen" />
	<!-- <node pkg="faster_lio" type="pointcloud2livox.py" name="livox_laser_transform" output="screen" /> -->

	<arg name="map" default="$(find faster_lio)/PCD/scans.pcd" />
	<!-- loalization-->
    <!-- <node pkg="fast_lio_localization" type="global_localization.py" name="global_localization" output="screen" /> -->

	<!-- transform  fusion-->
    <!-- <node pkg="fast_lio_localization" type="transform_fusion.py" name="transform_fusion" output="screen" /> -->

	<!-- glbal map-->
    <node pkg="pcl_ros" type="pcd_to_pointcloud" name="map_publishe" output="screen" args="$(arg map) 5 _frame_id:=world cloud_pcd:=map" />
    <node pkg="map_server" type="map_server" name="map_server" args="$(find faster_lio)/maps/map.yaml">
		<param name="frame_id" value="world" />
		<remap from="/map" to="/map2d" />
	</node>


	<node pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" name="pointcloud_to_laserscan">
		<remap from="cloud_in" to="/cur_scan_in_map"/>
		<!--
			参数解释如下：

			use_inf: 是否启用无穷远距离（inf）输出。true表示障碍物外的距离将填充值为无穷大。
			inf_epsilon: 定义无穷大距离的数值，实际输出的最大float值。默认为1.0。
			concurrency_level: 并发线程数。1表示单线程运行节点，也可以提高以加速点云到激光的转换速度。
			angle_increment: 激光扫描中每个角度步进，单位为弧度（rad），0.0087约等于0.5度。
			scan_time: 一帧激光的扫描时间，单位秒。这里设置为10秒，并不是激光工作频率，仅用作消息时间戳填充，对动态环境扫描影响较小。
			range_min: 输出的激光的最小距离阈值，单位米。小于此值的数据会被过滤。
			range_max: 输出的激光的最大距离阈值，单位米。大于此值的数据会被过滤。
			transform_tolerance: 坐标变换的容忍延迟，单位秒，用于tf坐标查询超时时间。
			min_height: 在点云中用于投影到激光平面时，允许的点的最小高度（z），单位米。低于此高度的点会被滤除。
			max_height: 同理，允许点的最大高度，单位米。高于此高度的点不会被用于生成激光数据。
			angle_min: 激光输出的起始角度，单位为弧度（-1.5708为-90度）。
			angle_max: 激光输出的终止角度，单位为弧度（1.5708为+90度）。
		-->
		<param name="use_inf" value="true" />
		<param name="inf_epsilon" value="1.0" />
		<param name="concurrency_level" value="1" />
		<param name="angle_increment" value="0.004" />
		<param name="scan_time" value="0.04" />
		<param name="range_min" value="0.3" />
		<param name="range_max" value="30.0" />
		<param name="transform_tolerance" value="0.01" />
		<param name="min_height" value="-1.4" />
		<param name="max_height" value="0.1" />
		<param name="angle_min" value="-3.14159" />
		<param name="angle_max" value="3.14159" />
	</node>

    <node pkg="move_base" type="move_base" name="move_base">
        <rosparam file="$(find faster_lio)/nav_lidar/costmap_common_params.yaml" command="load" ns="global_costmap" />
        <rosparam file="$(find faster_lio)/nav_lidar/costmap_common_params.yaml" command="load" ns="local_costmap" />
        <rosparam file="$(find faster_lio)/nav_lidar/global_costmap_params.yaml" command="load" />
        <rosparam file="$(find faster_lio)/nav_lidar/local_costmap_params.yaml" command="load" />
        <rosparam file="$(find faster_lio)/nav_lidar/local_planner_params.yaml" command="load" /> <!-- 加载 local planner 参数 -->
        <param name="base_global_planner" value="global_planner/GlobalPlanner" /> 
        <!-- <param name="base_local_planner" value="base_local_planner/TrajectoryPlannerROS" /> -->
	<!-- <param name="base_local_planner" value="wpbh_local_planner/WpbhLocalPlanner" /> -->
	<param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
        <remap from="/map" to="/map2d" />
        <remap from="odom" to="/localization" />
	<!-- move_base 对外发布的速度话题（DWA 算出来的速度最终由 move_base 发布） -->
		<remap from="cmd_vel" to="$(arg cmd_vel_topic)" />

    </node>

	<group if="$(arg rviz)">
		<node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find faster_lio)/config/rviz/faster_lio_loc.rviz" />
	</group>

</launch>